ocaml/xapi/console.ml

```diff
@@ -84,26 +112,44 @@ let address_of_console __context console : address option =
     ) ;
   address_option
 
-let real_proxy __context _ _ vnc_port s =
-  try
-    Http_svr.headers s (Http.http_200_ok ()) ;
-    let vnc_sock =
-      match vnc_port with
-      | Port x ->
-          Unixext.open_connection_fd "127.0.0.1" x
-      | Path x ->
-          Unixext.open_connection_unix_fd x
-    in
-    (* Unixext.proxy closes fds itself so we must dup here *)
-    let s' = Unix.dup s in
-    debug "Connected; running proxy (between fds: %d and %d)"
-      (Unixext.int_of_file_descr vnc_sock)
-      (Unixext.int_of_file_descr s') ;
-    Unixext.proxy vnc_sock s' ;
-    debug "Proxy exited"
-  with exn -> debug "error: %s" (ExnHelper.string_of_exn exn)
-
-let ws_proxy __context req protocol address s =
+let real_proxy __context vm _ _ vnc_port s =
+  let vm_id = Ref.string_of vm in
+  let pool = Helpers.get_pool ~__context in
+  let is_limit_set = Db.Pool.get_limit_console_sessions ~__context ~self:pool in
+  if is_limit_set && not (Connection_limit.try_add vm_id) then
```

Comment:
Similarly, you can:
rename `real_proxy` to `real_proxy'` and update its parameters; and
do something like:
```
let real_proxy __context vm _ _ vnc_port s =
  let f () = real_proxy' vnc_port s in
  let pool = Helpers.get_pool ~__context in
  if Db.Pool.get_limit_console_sessions ~__context ~self:pool then
    with_limit_check vm f
  else
    f ()
```

Reply:
And 
```
let with_limit_check vm f =
  if try_add vm then
    finnaly
      (f ())
      (remove vm)
  else
    Http_svr.headers s (Http.http_503_service_unavailable ())
```